core/services/keeper/upkeep_executer.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/atomic"
+	"gorm.io/gorm"
+)
+
+const (
+	checkUpkeep        = "checkUpkeep"
+	performUpkeep      = "performUpkeep"
+	executionQueueSize = 10
+)
+
+func NewUpkeepExecuter(
+	db *gorm.DB,
+	ethClient eth.Client,
+) UpkeepExecuter {
+	return UpkeepExecuter{
+		blockHeight:    atomic.NewInt64(0),
+		ethClient:      ethClient,
+		keeperORM:      NewORM(db),
+		isRunning:      atomic.NewBool(false),
+		executionQueue: make(chan struct{}, executionQueueSize),
+		chDone:         make(chan struct{}),
+		chSignalRun:    make(chan struct{}, 1),
+	}
+}
+
+// UpkeepExecuter fulfills HeadTrackable interface
+var _ store.HeadTrackable = UpkeepExecuter{}
+
+type UpkeepExecuter struct {
+	blockHeight *atomic.Int64
+	ethClient   eth.Client
+	keeperORM   KeeperORM
+	isRunning   *atomic.Bool
+
+	executionQueue chan struct{}
+	chDone         chan struct{}
+	chSignalRun    chan struct{}
+}
+
+func (executer UpkeepExecuter) Start() error {
+	if executer.isRunning.Load() {
+		return errors.New("already started")
+	}
+	executer.isRunning.Store(true)
+	go executer.run()
+	return nil
+}
+
+func (executer UpkeepExecuter) Close() error {
+	close(executer.chDone)
+	return nil
+}
+
+func (executer UpkeepExecuter) Connect(head *models.Head) error {
+	return nil
+}
+
+func (executer UpkeepExecuter) Disconnect() {}
+
+func (executer UpkeepExecuter) OnNewLongestChain(ctx context.Context, head models.Head) {
+	executer.blockHeight.Store(head.Number)
+	// avoid blocking if signal already in buffer
+	select {
+	case executer.chSignalRun <- struct{}{}:
+	default:
+	}
+}
+
+func (executer UpkeepExecuter) run() {
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case <-executer.chSignalRun:
+			executer.processActiveUpkeeps()
+		}
+	}
+}
+
+func (executer UpkeepExecuter) processActiveUpkeeps() {
+	// Keepers could miss their turn in the turn taking algo if they are too overloaded
+	// with work because processActiveUpkeeps() blocks
+	logger.Debug("received new block, running checkUpkeep for keeper registrations")
+
+	activeUpkeepss, err := executer.keeperORM.EligibleUpkeeps(executer.blockHeight.Load())
+	if err != nil {
+		logger.Errorf("unable to load active registrations: %v", err)
+		return
+	}
+
+	for _, reg := range activeUpkeepss {
+		executer.concurrentExecute(reg)
+	}
+}
+
+func (executer UpkeepExecuter) concurrentExecute(upkeep UpkeepRegistration) {
+	executer.executionQueue <- struct{}{}
+	go executer.execute(upkeep)
+}
+
+// execute will call checkForUpkeep and, if it succeeds, triger a job on the CL node
+// DEV: must perform contract call "manually" because abigen wrapper can only send tx
+func (executer UpkeepExecuter) execute(upkeep UpkeepRegistration) {
+	defer func() { <-executer.executionQueue }()
```

Comment:
This is an interesting pattern... what does it do?

Reply:
It sets an upper limit on the number of concurrent executions (currently a constant in the file, but have a story to make it configurable). Basically if we have 1000 upkeeps, rather than spin up 1000 goroutines, just spin up 10 and and start a new one once the old one finishes.

Reply:
Not glued to this pattern if you think there's a better way, BTW!